cpp/src/Slice/Parser.cpp

```diff
@@ -522,47 +522,88 @@ Slice::DocComment::parseFrom(const ContainedPtr& p, DocLinkFormatter linkFormatt
 
         if (parseNamedCommentLine(l, paramTag, name, lineText))
         {
-            if (!isOperation)
+            if (!operationTarget)
             {
                 // If '@param' was put on anything other than an operation, ignore it and issue a warning.
-                string msg = "the '" + paramTag + "' tag is only valid on operations";
+                const string msg = "the '" + paramTag + "' tag is only valid on operations";
                 p->unit()->warning(p->file(), p->line(), InvalidComment, msg);
                 currentSection = nullptr;
             }
             else
             {
-                comment._parameters[name] = {};
-                currentSection = &(comment._parameters[name]);
-            }
-        }
-        else if (parseNamedCommentLine(l, throwsTag, name, lineText))
-        {
-            if (!isOperation)
-            {
-                // If '@throws' was put on anything other than an operation, ignore it and issue a warning.
-                string msg = "the '" + throwsTag + "' tag is only valid on operations";
-                p->unit()->warning(p->file(), p->line(), InvalidComment, msg);
-                currentSection = nullptr;
-            }
-            else
-            {
-                comment._exceptions[name] = {};
-                currentSection = &(comment._exceptions[name]);
+                // Check that the '@param <name>' corresponds to an actual parameter in the operation.
+                const ParameterList params = operationTarget->parameters();
+                const auto paramNameCheck = [&name](const ParameterPtr& param) { return param->name() == name; };
+                if (std::find_if(params.begin(), params.end(), paramNameCheck) == params.end())
+                {
+                    const string msg = "'" + paramTag + " " + name +
+                                       "' does not correspond to any parameter in operation '" + p->name() + "'";
+                    p->unit()->warning(p->file(), p->line(), InvalidComment, msg);
+                }
+
+                // Check if this is a duplicate tag. If it is, ignore it and issue a warning.
+                if (comment._parameters.count(name) != 0)
+                {
+                    const string msg = "ignoring duplicate doc-comment tag: '" + paramTag + " " + name + "'";
+                    p->unit()->warning(p->file(), p->line(), InvalidComment, msg);
+                    currentSection = nullptr;
+                }
+                else
+                {
+                    comment._parameters[name] = {};
+                    currentSection = &(comment._parameters[name]);
+                }
             }
         }
-        else if (parseNamedCommentLine(l, exceptionTag, name, lineText))
+        else if (
+            parseNamedCommentLine(l, throwsTag, name, lineText) ||
+            parseNamedCommentLine(l, exceptionTag, name, lineText))
```

Comment:
We support `@exception` as an alias for `@throws`, and used to parse/validate them separately.
I combined them both into a single block to avoid code-duplication.